buildbot_nix/buildbot_nix/nix_eval.py

```diff
@@ -0,0 +1,639 @@
+            props.setProperty("virtual_builder_name", effect, source)
+            props.setProperty("status_name", f"effects.{effect}", source)
+            props.setProperty("virtual_builder_tags", "", source)
+            props.setProperty("command", effect, source)
+
+            triggered_schedulers.append((self.effects_scheduler, props))
+
+        return triggered_schedulers
+
+
+def nix_eval_config(
+    project: GitProject,
+    worker_names: list[str],
+    git_url: str,
+    nix_eval_config: NixEvalConfig,
+    build_config: Any,  # BuildConfig from nix_build module
+) -> BuilderConfig:
+    """Uses nix-eval-jobs to evaluate hydraJobs from flake.nix in parallel.
+    For each evaluated attribute a new build pipeline is started.
+    """
+    factory = util.BuildFactory()
+    # check out the source
+    url_with_secret = util.Interpolate(git_url)
+    factory.addStep(
+        GitLocalPrMerge(
+            repourl=url_with_secret,
+            method="clean",
+            submodules=True,
+            haltOnFailure=True,
+            logEnviron=False,
+            sshPrivateKey=project.private_key_path.read_text()
+            if project.private_key_path
+            else None,
+            sshKnownHosts=project.known_hosts_path.read_text()
+            if project.known_hosts_path
+            else None,
+        ),
+    )
+    drv_gcroots_dir = util.Interpolate(
+        f"/nix/var/nix/gcroots/per-user/{nix_eval_config.gcroots_user}/%(prop:project)s/drvs/",
+    )
+
+    factory.addStep(
+        NixEvalCommand(
+            project=project,
+            env={"CLICOLOR_FORCE": "1"},
+            name="Evaluate flake",
+            nix_eval_config=nix_eval_config,
+            haltOnFailure=True,
+            locks=[nix_eval_config.eval_lock.access("exclusive")],
+            drv_gcroots_dir=drv_gcroots_dir,
+            logEnviron=False,
+        ),
+    )
+
+    # Register gcroots for all skipped builds at once
+    factory.addStep(
+        RegisterSkippedGcroots(
+            project=project,
+            gcroots_user=nix_eval_config.gcroots_user,
+            branch_config=build_config.branch_config_dict,
+            name="Register gcroots for skipped builds",
+            doStepIf=lambda s: s.getProperty("event") == "push"
+            and build_config.branch_config_dict.do_register_gcroot(
+                project.default_branch, s.getProperty("branch")
+            ),
+            hideStepIf=lambda results, _s: results == util.SKIPPED,
+        ),
+    )
+    factory.addStep(
+        steps.ShellCommand(
+            name="Cleanup drv paths",
+            command=[
+                "rm",
+                "-rf",
+                drv_gcroots_dir,
+            ],
+            logEnviron=False,
+        ),
+    )
+    factory.addStep(
+        BuildbotEffectsCommand(
+            project=project,
+            env={},
+            name="Evaluate effects",
+            command=[
+                "buildbot-effects",
```

Comment:
https://buildbot.thalheim.io/#/builders/105/builds/302/steps/6/logs/stdio

I noticed that repo doesn't seem work (not actually sure if it ever worked). rev, branch, repo all seem to handled automatically by buildbot-effects, is there reasons to set them explicitly?

Reply:
Can you make an issue for this?